rfcs/0000-terraform-workflow-improvements.md

```diff
@@ -0,0 +1,826 @@
+
+#### Proposed Module Names
+
+With all the above in mind, the following are the proposed names for all existing and future root modules. (Note that the naming convention for child modules is different, and is covered in the [module versioning][guide-module-versioning] section.)
+
+Root modules associated with the "CTD Production" AWS account that should be renamed:
+
+- `bootstrap` → `aws-ctd-main-bootstrap`
+- `global` → `aws-ctd-main-base`
+- `dev` → multiple modules:
+  - move application development/staging resources to team-based modules:
+    - `aws-ctd-main-team-dataplatform`
+    - `aws-ctd-main-team-glides`
+    - `aws-ctd-main-team-lamp`
+    - `aws-ctd-main-team-pass-programs`
+    - `aws-ctd-main-team-screens`
+    - `aws-ctd-main-team-trc`
+    - `aws-ctd-main-team-website`
+  - move base dev resources to `aws-ctd-main-base`
+- `onprem_dev` → `aws-onprem-dev`
+- `onprem_prod` → `aws-onprem-prod`
+- `prod` → `aws-ctd-main-apps` (to remain as a single module for now)
+- `restricted` → `aws-ctd-main-restricted`
+
+Root modules to remain as-is:
+
+- `scalr`
+
+New root modules for "mbta-ctd-test" AWS account:
+
+- `aws-ctd-test-base`
+- `aws-ctd-test-bootstrap`
+- `aws-ctd-test-restricted`
+
+Possible future root modules:
+
+- `github`
+- `sentry`
+- `splunk`
+
+## 2. Module Testing, Detailed
+[reference-module-testing]: #2-module-testing-detailed
+
+As noted in the [guide-level explanation][guide-module-testing], we will set up testing for each application and infrastructure component child module by creating a `test/` module within each module's directory. The test module will define any dependent resources required by the module, and then call the module itself. It could even call other application modules if needed.
+
+### Testing Workflow
+
+To use a module's `test/` module for testing:
+
+1. On a new feature branch, make modifications to the functionality of the module
+1. In that same branch, add test cases for the new functionality to the module's testing configuration in `test/`
+1. Switch to the default branch (`master`/`main`), `cd` to the `test/` directory and run `terraform apply` to stand up the test resources
+1. Switch to the feature branch and run `terraform plan`/`terraform apply` to see the impact of the changes on the pre-existing resources
+1. When finished testing, run `terraform destroy` to tear down all test resources
+
+This workflow is intended to be carried out manually at first, but we might be able to use GitHub Actions to automate it at some point.
```

Comment:
when using this workflow i mostly enjoyed how quickly i could get feedback on the changes i wanted to make. one thing to note though is that standing up databases can take quite a bit of time. it wasn't an end of the world problem, but if a project uses an RDS and is standing up multiple tests, the automated tests could stretch out.